File: leetcode/max-points-on-a-line/max-points-on-a-line.cpp
/**************************************
* @author dujiong
* @date 2016.11.22
* @version V0.1
**************************************/

/**
 * Definition for a point.
 * struct Point {
 *     int x;
 *     int y;
 *     Point() : x(0), y(0) {}
 *     Point(int a, int b) : x(a), y(b) {}
 * };
 */
class Solution {
public:
    int maxPoints(vector<Point> &points)
	{
    	int size = points.size();
    	if(size == 0 || size == 1)
        	return size;
 
    	int ret = 0;
    	for(int i = 0; i < size; ++i)
    	{
        	int curmax = 1;
        	map<double,int> mp;
        	int vcnt = 0; 	//垂直点
        	int dup = 0; 	//重复点
        	for(int j = 0; j < size; ++j)
       	 	{
            	if(j != i)
            	{
                	double x1 = points[i].x - points[j].x;
                	double y1 = points[i].y - points[j].y;
                	if(x1 == 0 && y1 == 0)		//和选定的起始点重复的点
                    	dup++;
                	else
                	{
                    	double k = x1 == 0 ? INT_MAX : y1/x1;
                    	if(mp[k] == 0)
                        	mp[k] = 2;
                    	else
                        	++mp[k];
                    	curmax = max(mp[k], curmax);			
                	}                
            	}
        	}
        	ret = max(ret, curmax+dup);      		
    	}
    	return ret;
 	}
};

File: leetcode-golang/Max Points on a Line/Max Points on a Line.go
func maxPoints(points [][]int) int {
	if len(points) == 0 || len(points[0]) == 0 {
		return 0
	}
	res := 0
	for i := 0; i < len(points); i++ {
		duplicate := 1
		for j := i + 1; j < len(points); j++ {
			cnt := 0
			x1, y1 := points[i][0], points[i][1]
			x2, y2 := points[j][0], points[j][1]
			if x1 == x2 && y1 == y2 {
				duplicate++
				continue
			}
			for k := 0; k < len(points); k++ {
				x3, y3 := points[k][0], points[k][1]
				//行列式
				if x1*y2+x2*y3+x3*y1-x3*y2-x2*y1-x1*y3 == 0 {
					cnt++
				}
			}
			if res < cnt {
				res = cnt
			}
		}
		if res < duplicate {
			res = duplicate
		}
	}
	return res
}